xla/service/gpu/nccl_collective_thunk.cc

```diff
@@ -224,6 +224,49 @@ StatusOr<std::vector<DeviceBufferPair>> ConvertToDeviceBuffers(
   return device_buffers;
 }
 
+Status MaybeRegisterBuffers(int device_ordinal,
+                            const std::vector<DeviceBufferPair>& buffers,
+                            ncclComm_t comm) {
+  // TODO(tmorris): Only register if xla_gpu_enable_nccl_user_buffers is true.
+#if XLA_ENABLE_XCCL
+  // Keep track of which communicators we have registered for already.
+  // Each device has one NCCL buffer which only needs to be registered once per
+  // each comm.
+  struct RegisteredBuffers {
```

Comment:
Nit: probably more readable to have the struct outside, in an anonymous namespace.